### PR TITLE
add business logic for friends recommendations BE

### PR DESCRIPTION
**Pull request by rostyk1331**

### CODE REVIEW

File: server/src/helpers/getFriendsRecommendations.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            if (usersFriendshipDegree[friend] === Infinity) {
+                usersFriendshipDegree[friend] = usersFriendshipDegree[current] + 1;
+                queue.push(friend);
+            }
+        });
+    }
+    const sortedListOfRecommendations = Object.keys(friendsRecommendations)
+          .sort((a: string, b: string) => friendsRecommendations[b] - friendsRecommendations[a]);
+
+    return sortedListOfRecommendations;
```
Comment (InnaKutsik) [inline server/src/helpers/getFriendsRecommendations.ts]: split into function. It is too long one.

File: server/src/db.requests/getFriendsRecommendations.requests.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    } catch (e) {
+        next({status: 409, message: 'User does not exist'});
+    }
+};
+
+export const findUsers = async (listOfUsers: string[]): Promise<object[]> => {
+    const users = await User.find({_id: {$in: listOfUsers}})
+        .select('username photoPath');
+
+    users.sort((a: any, b: any) => listOfUsers.findIndex((id: string) => a._id.equals(id)) -
```
[the remark this answers is not in the document]
Comment (rostyk1331) [reply server/src/db.requests/getFriendsRecommendations.requests.ts]: Fixed.

File: server/src/helpers/getFriendsRecommendations.ts (hunk 1 above)
Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Fixed.

File: server/src/helpers/getFriendsRecommendations.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    graph[user._id] = user.following.map((fol: any) => fol._id);
+
+    user.following.forEach((fol: any) => {
+        graph[fol._id] = fol.following.map((f: object) => f.toString());
+        fol.following.forEach((f: string) => {
+            graph[f] = [];
+        });
+    });
+
+    return graph;
```
[the remark this answers is not in the document]
Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Is it better?

[the remark this answers is not in the document]
Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Fixed.

File: server/src/helpers/getFriendsRecommendations.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+            if (usersFriendshipDegree[friend] === Infinity) {
+                usersFriendshipDegree[friend] = usersFriendshipDegree[current] + 1;
+                queue.push(friend);
+            }
+        });
+    }
+
+    return friendsRecommendations;
+};
```
Comment (InnaKutsik) [inline server/src/helpers/getFriendsRecommendations.ts]: split into function. It is too long one.

File: server/src/helpers/getFriendsRecommendations.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                return {...res, [follow]: [] };
+            }
+            return res;
+        }, result);
+
+        return {...updatedByFollowingResult, [fol._id]: fol.following.map((f: object) => f.toString())};
+    }, graph);
+};
+
+export const findRecommendations = (users: IGraph, root: string): IFriendsRecommendations => {
```
Comment (InnaKutsik) [inline server/src/helpers/getFriendsRecommendations.ts]: root - what is it?

File: server/src/helpers/getFriendsRecommendations.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    const usersFriendshipDegree: IUsersFriendshipDegree = {};
+    const friendsRecommendations: IFriendsRecommendations = {};
+
+    const allUsers = Object.keys(users);
+    allUsers.forEach((user: string) => usersFriendshipDegree[user] = Infinity);
```
Comment (InnaKutsik) [inline server/src/helpers/getFriendsRecommendations.ts]: why Infinity ?

Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Add comment to code. Infinity here means that node is have not been visited yet.

File: server/src/helpers/getFriendsRecommendations.ts (hunk 4 above)
Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Done.

File: server/src/helpers/getFriendsRecommendations.ts (hunk 5 above)
Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Add comment to code. It is starting node, from which we start traversing the graph (logged user, for which we are finding friends recommendations).

File: server/src/helpers/getFriendsRecommendations.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    // (in our case, distance 1 - they are friends, 2 - have mutual friend, and so on)
+    const usersFriendshipDegree: IUsersFriendshipDegree = {};
+    // Object, where key - user to be recommended, value - recommendation score
+    const friendsRecommendations: IFriendsRecommendations = {};
+
+    // Take all nodes of graph and set distance to rootUser to Infinity, which means, that node is not visited yet
+    const allUsers = Object.keys(users);
+    allUsers.forEach((user: string) => usersFriendshipDegree[user] = Infinity);
+    // Set distance from root node to itself to 0
+    usersFriendshipDegree[rootUser] = 0;
```
Comment (InnaKutsik) [inline server/src/helpers/getFriendsRecommendations.ts]:
```
const startFriendshipDegreeObj = { [rootUser]: 0};
const usersFriendshipDegree: IUsersFriendshipDegree = Object.keys(users).reduce((result, item) => ({...result, [item]: Infinity }), startFriendshipDegreeObj);
const friendsRecommendations: IFriendsRecommendations = {};

Comment (rostyk1331) [reply server/src/helpers/getFriendsRecommendations.ts]: Fixed.

**Pull request merged**